Add a deprecatedPropType module to show deprecation warnings

janicduplessis opened the pull request:
To allow smoother API changes for users we often deprecate props and keep them around for a while before removing them. Right now it is all done manually, this adds a consistent way to show a warning when using a deprecated prop.

This also adds a deprecation warning of the website generated from the deprecatedPropType.

<img width="643" alt="screen shot 2016-01-26 at 7 43 08 pm" src="https://cloud.githubusercontent.com/assets/2677334/12600172/7af28fb0-c465-11e5-85e5-3786852bf522.png">

It also changes places where we added the warnings manually to use deprecatedPropType instead.

janicduplessis commented:
The only issue there is right now is that it generates 2 warnings, 1 for the JS component and 1 for the native one, I didn't really find a clean way to fix this since there is no way to make the difference between a native and normal component. One hacky way to do it would be to check if the component name starts with 'RCT' or add a prop on the propType object thats says it is a native component.

satya164 commented:
So awesome. Thanks @janicduplessis 

Checking `RCT` prefix won't work for many native components.

cc @mkonicek @nicklockwood

janicduplessis commented on Libraries/Utilities/deprecatedPropType.js:
```diff
@@ -0,0 +1,30 @@
+function deprecatedPropType(
+  propType: ReactPropsCheckType,
+  explanation: string
+): ReactPropsCheckType {
```
Should we add this to eslint globals or is there a way to import these types?

vjeux replied:
Add it to eslint global

janicduplessis replied:
Fixed in #5585

vjeux replied:
What I thought but wanted to double check :)

satya164 commented:
I don't know where this type is defined. If it's in a file included via `.flowconfig`, we should define it as global. Otherwise `import type ReactPropsCheckType from './path'`

janicduplessis commented:
@satya164 It's a type from react so it's builtin in flow. I guess the only way is to add it as a global in eslint.

satya164 commented:
@janicduplessis Yup!

janicduplessis commented:
@satya164 I'll make a PR soon to add all flow global types to eslint so the bot stop complaining :)

janicduplessis commented:
Ok I figured a clean way to check if it's a native module, I can just check if UIManager contains the component name.

nicklockwood commented:
This looks great to me, but I think @vjeux is probably the best person to review it.

vjeux commented on website/server/docgenHelpers.js:
```diff
@@ -34,6 +35,43 @@ function stylePropTypeHandler(documentation, path) {
   });
 }
 
+function deprecatedPropTypeHandler(documentation, path) {
```
So good

vjeux commented on website/layout/AutodocsLayout.js:
```diff
@@ -108,6 +108,15 @@ var ComponentDoc = React.createClass({
         </Header>
+        {prop.deprecationMessage && <div className="deprecated">
+          <div className="deprecatedTitle">
+            <img className="deprecatedIcon" src="/react-native/img/Warning.png" />
```
Was the image already there? I don't see it in your commit

janicduplessis replied:
Yes it's there in the first of the 2 commits at the bottom.

vjeux commented on Libraries/Utilities/deprecatedPropType.js:
```diff
@@ -0,0 +1,33 @@
+  propType: ReactPropsCheckType,
+  explanation: string
+): ReactPropsCheckType {
+  return function validate(props, propName, componentName) {
```
I assume that this function is only going go be called when dev, but could you double check

janicduplessis replied:
Double checked and this doesn't get called when __DEV__ is false :)

vjeux commented:
This is awesome, thanks for building this. Going to make our upgrade path much smoother :)

vjeux commented:
@facebook-github-bot shipit

janicduplessis commented:
Damn @vjeux in a 1300 lines css file we managed to both add new styles at the exact same place in the middle of the file :p

Rebased and fixed the conflict.

vjeux commented:
Les grands esprits se rencontrent ;)

vjeux commented:
@facebook-github-bot shipit

janicduplessis commented:
Haha!
